enhancements/coreos-encrypted-by-default.md

```diff
@@ -0,0 +1,425 @@
+
+## Terms
+* LUKS/LUKS2: the standard Linux encryption method. Encrypted filesystems stored in a container.
+* Master Key: the _actual_ encryption key used. The master key encrypted using a passphrase and is embedded in the LUKS meta-data on disk encrypted. This encryption key is unknown to users.
+* LUKS. Passphrase: any string or file that can be used to unlock the master key. Each passphrase is bound to a "key-slot."
+* LUKS Token: arbitrary JSON stored in the LUKS meta-data.
+* NBDE: [Network Bound-Disk Encryption](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Security_Guide/sec-Using_Network-Bound_Disk_Encryption.html). A method of providing context-based disk encryption.
+* [Clevis](https://github.com/latchset/clevis): Linux CLI tools for policy-based encryption and decryption. Clevis has a Dracut module for automatically opening LUKS containers in an NBDE context.
+* Clevis pin: a program called by Clevis to encrypt and decryption of secrets.
+* Tang: A server implementation that provides cryptographic binding services without the need for an escrow. Secrets are derived from what the client and the Tang server both have, reducing the attack surface.
+* KMS: Key Management System, such as Amazon KMS.
+* [TPM2](https://en.wikipedia.org/wiki/Trusted_Platform_Module): A secure cryptoprocessor for storing keys in hardware.
+* Dracut Module: program(s) that run early in boot before the on-disk Linux is booted.
+
+If you are unfamiliar with some of these terms [please see an excellent Youtube video on Clevis/Tang](https://youtu.be/2uLKvB8Z5D0).
+
+## Proposal
+
+To provide install-time _and_ rolling re-encryption, CoreOS boot images will:
+* Have the root filesystem housed in a LUKS container.
+* Use the 'null-cipher' for the initial LUKS container "encryption". This allows _any_ passphrase to unlocked the LUKS container.
+* The passphrase will be stored in a LUKS token via the Clevis plaintext pin.
+
+Note: the default boot image is _effectively *NOT* encrypted_.
+
+On first boot, a Dracut module will re-encrypt the LUKS container unless the kernel CLI parameter ('rd.coreos.no_encryption=1') is used.  The module will likely run after Ignition has completed. It will:
+  * Prepare a random passphrase,
+  * Randomize the LUKS container's UUID,
+  * Re-encrypt the LUKS container using AES-256-CBC (FIPS 140-2 compliant). This process generates a new random master key.
+  * If a Clevis configration is found, apply it in the initramfs. Otherwise:
+      * if a TPM2 device is found, bind the LUKS container to TPM2.
+      * if no TPM2 device, use the plaintext Clevis pin
+
+
+The Dracut modules will also allow both first-boot and reboot re-encryption using kernel commandline arguments. The re-encryption step is necessary to ensure that the LUKS master key is unique to each instance of CoreOS. In testing, the re-encryption step takes between 1 and 3 minutes, depending on the CPU and the backing disk-type.
```

Comment:
> In testing, the re-encryption step takes between 1 and 3 minutes, depending on the CPU and the backing disk-type.

Can we provide more results using cloud instance types and disks? it would be at least useful to provide the results for the default types being used in OpenShift today.

Reply:
I like this idea! I'd suggest adding this information into the resulting documentation due to the time left for these enhancement proposals.